core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java

```diff
@@ -196,46 +195,54 @@ final class ListeningFuture extends CompletableFuture<Endpoint> implements Consu
             this.executor = executor;
         }
 
-        @Override
-        public void accept(Void updated) {
+        /**
+         * Returns {@code true} if an {@link Endpoint} has been selected.
+         */
+        public boolean tryComplete() {
             if (selectedEndpoint != null || isDone()) {
-                return;
+                return true;
             }
 
             try {
                 final Endpoint endpoint = selectNow(ctx);
-                if (endpoint != null) {
-                    cleanup();
-
-                    // Complete with the selected endpoint.
-                    selectedEndpoint = endpoint;
-                    executor.execute(() -> super.complete(endpoint));
+                if (endpoint == null) {
+                    return false;
                 }
+
+                cleanup(false);
+                // Complete with the selected endpoint.
+                selectedEndpoint = endpoint;
+                executor.execute(() -> super.complete(endpoint));
```

Comment:
Question) If a `ListeningFuture` is adding to `pendingFutures` and completed by `tryComplete`, will this future ever be removed from `pendingFutures`?

It seems like `ListeningFuture.complete(endpoint)` is never called, and hence `cleanup(true)` is never called

Reply:
> If a ListeningFuture is adding to pendingFutures and completed by tryComplete, will this future ever be removed from pendingFutures?

If `tryComplete()` returns true, the `ListeningFuture` will be removed from `pendingFutures` by `pendingFutures.removeIf()`.

Reply:
Ah, I missed that point. Looks good then 👍